### PR TITLE
chore: bump supported tutor versions to work on tutor 16 for palm

### PR DESCRIPTION
**Pull request by xitij2000**
This plugin works perfectly with the latest release of tutor as-is, so this change expands its compatibility to work with the latest relese.

### CODE REVIEW

File: .github/workflows/tox.yml [hunk 1]
```diff
@@ -12,6 +12,8 @@ jobs:
         python-version:
           - 3.8
           - 3.9
+          - 3.10
+          - 3.11
```
Comment (fghaas) [inline .github/workflows/tox.yml:16]: For this to not break the build, you'll have to quote the string "3.10" here.

File: setup.py [hunk 2]
```diff
@@ -24,7 +24,7 @@
     packages=find_packages(exclude=["tests*"]),
     include_package_data=True,
     python_requires=">=3.8",
-    install_requires=["tutor <16, >=14.0.0"],
+    install_requires=["tutor>=14.0.0"],
```
Comment (fghaas) [inline setup.py:27]: Please don't just drop the upper version bound, but instead replace `<16` with `<17`.

**Review by fghaas — COMMENTED** [review]
@xitij2000 Thanks for the PR! I've added a couple of inline comments, please address those.

Also, please add a one-line update to the README indicating that this plugin works with Palm now, and also add an entry to `CHANGELOG.md`, like so:

```markdown
## Unreleased

* [Enhancement] Support Tutor 16 and Open edX Palm, Python 3.10, and Python 3.11.
```

Thanks again!

**Review by kaustavb12 — APPROVED** [review]
:+1:

@xitij2000 I don't have anything to add here, other than the things already pointed out by @fghaas.
Since the changes are minor, I am approving this from my end, subject to the resolution of the review comments.

- [x] I tested this: Did not test these, since the changes are mostly version bumps
- [x] I read through the code
- [ ] Includes documentation

**Comment by xitij2000** [comment]
@fghaas  I've applied the changes you suggested. Thanks for the quick feedback!

**Comment by fghaas** [comment]
Excellent, thank you! The one thing that's still needed is adding a line for Palm in the compatibility matrix in the README.

**Comment by bradenmacdonald** [comment]
Guys, I just found that this is not compatible with Palm as is. It needs

```
COURSE_METADATA_EXPORT_STORAGE = DEFAULT_FILE_STORAGE
```

added to `tutors3/patches/openedx-cms-common-settings`

Otherwise, [tutor's `production.py`](https://github.com/overhangio/tutor/blob/master/tutor/templates/apps/openedx/settings/cms/production.py) first loads the platform's [`production.py`](https://github.com/openedx/edx-platform/blob/fa06be106eae56b7e6427f947949b32a5d544d16/cms/envs/production.py) which sets `DEFAULT_FILE_STORAGE = 'storages.backends.s3boto.S3BotoStorage'` (which is no longer actually included in the platform's specified version of django-storages), and then it sets `COURSE_METADATA_EXPORT_STORAGE = DEFAULT_FILE_STORAGE`.

So when you run an image built using this on Tutor 16, the LMS works but the CMS gives this error:

```
ModuleNotFoundError: No module named 'storages.backends.s3boto'
```

**Comment by xitij2000** [comment]
> Guys, I just found that this is not compatible with Palm as is. It needs
> 
> ```
> COURSE_METADATA_EXPORT_STORAGE = DEFAULT_FILE_STORAGE
> ```
> 
> added to `tutors3/patches/openedx-cms-common-settings`
> 
> Otherwise, [tutor's `production.py`](https://github.com/overhangio/tutor/blob/master/tutor/templates/apps/openedx/settings/cms/production.py) first loads the platform's [`production.py`](https://github.com/openedx/edx-platform/blob/fa06be106eae56b7e6427f947949b32a5d544d16/cms/envs/production.py) which sets `DEFAULT_FILE_STORAGE = 'storages.backends.s3boto.S3BotoStorage'` (which is no longer actually included in the platform's specified version of django-storages), and then it sets `COURSE_METADATA_EXPORT_STORAGE = DEFAULT_FILE_STORAGE`.
> 
> So when you run an image built using this on Tutor 16, the LMS works but the CMS gives this error:
> 
> ```
> ModuleNotFoundError: No module named 'storages.backends.s3boto'
> ```
> 
> I believe that the fix I specified above will correctly use `storages.backends.s3boto3.S3Boto3Storage` which is the right version and is included in the current django-storages.
> 
> We can also fix this upstream (the default shouldn't be a deprecated and uninstalled storage backend), but that might take a bit longer.

@bradenmacdonald 
This is kind of surprising, since I tested this and the CMS is definitely starting for me. Is it under specific conditions that this is happening? I have a few palm deployment running with this patch, and I haven't faced any problems there. 

However, I've also only deployed all of them along with our grove plugin, so perhaps that is masking this issue?

**Comment by fghaas** [comment]
> Guys, I just found that this is not compatible with Palm as is. It needs
> 
> ```
> COURSE_METADATA_EXPORT_STORAGE = DEFAULT_FILE_STORAGE
> ```
> added to `tutors3/patches/openedx-cms-common-settings`

Hmmm, OK. It would be good if we got to the bottom of this.

* If this change is needed for Palm, lest things break on that release, then this takes the patch category from `chore` to `fix`.
* If this fix is needed _only_ for Palm, and applying it were to break things on Olive and earlier, then we also need to set up a separate `olive` branch while we continue to track the latest release (i.e. currently Palm) in `main`.

I'm not sure if we'll get round to testing this today, but in case you do, we'd greatly appreciate if you shared your results here. :)

**Comment by bradenmacdonald** [comment]
Unfortunately I won't have time to dig into this further today. Technically the build I was testing on was slightly past Palm (on master) so that may be the issue. But I checked and that setting is still there on the Palm branch so I assumed it would also apply to Palm. Perhaps it's not actually affecting Palm or this only happens when interacting with some other aspect of our k8s deployment; I can confirm on Monday.

**Comment by fghaas** [comment]
Folks, if you have any updates on this one we'd very much appreciate them. :)

**Comment by bradenmacdonald** [comment]
OK, sorry for the false alarm. It turns out that with this plugin in Palm, `settings.COURSE_METADATA_EXPORT_STORAGE` _is_ incorrectly set to `storages.backends.s3boto.S3BotoStorage`, unlike `settings.DEFAULT_FILE_STORAGE` and all the other `_STORAGE` settings, which are correctly set to `boto3.S3Boto3Storage`.

However, Palm is currently using `django-storages==1.8` which still has the old boto storage backend included, so the only result is a warning in the console:

```
2023-07-19 18:30:49,235 WARNING 62 [py.warnings] [user None] [ip None] warnings.py:109 - /openedx/venv/lib/python3.8/site-packages/storages/backends/s3boto.py:41: DeprecationWarning: The S3BotoStorage backend is deprecated in favor of the S3Boto3Storage backend and will be removed in django-storages 1.8. This backend is mostly in bugfix only mode and has been for quite a while (in much the same way as its underlying library 'boto'). For performance, security and new feature reasons it is _strongly_ recommended that you update to the S3Boto3Storage backend. Please see the migration docs https://django-storages.readthedocs.io/en/latest/backends/amazon-S3.html#migrating-boto-to-boto3.
```

**However**, post-Palm, this PR https://github.com/openedx/edx-platform/pull/32350 has upgraded Django storages, so if you use any platform version since June 5, you'll get this error. Ultimately though as I said, I think the bug is an upstream bug so I'm going to open a PR to fix it upstream.

**Comment by xitij2000** [comment]
> OK, sorry for the false alarm. It turns out that with this plugin in Palm, `settings.COURSE_METADATA_EXPORT_STORAGE` _is_ incorrectly set to `storages.backends.s3boto.S3BotoStorage`, unlike `settings.DEFAULT_FILE_STORAGE` and all the other `_STORAGE` settings, which are correctly set to `boto3.S3Boto3Storage`.
> 
> However, Palm is currently using `django-storages==1.8` which still has the old boto storage backend included, so the only result is a warning in the console:
> 
> ```
> 2023-07-19 18:30:49,235 WARNING 62 [py.warnings] [user None] [ip None] warnings.py:109 - /openedx/venv/lib/python3.8/site-packages/storages/backends/s3boto.py:41: DeprecationWarning: The S3BotoStorage backend is deprecated in favor of the S3Boto3Storage backend and will be removed in django-storages 1.8. This backend is mostly in bugfix only mode and has been for quite a while (in much the same way as its underlying library 'boto'). For performance, security and new feature reasons it is _strongly_ recommended that you update to the S3Boto3Storage backend. Please see the migration docs https://django-storages.readthedocs.io/en/latest/backends/amazon-S3.html#migrating-boto-to-boto3.
> ```
> 
> **However**, post-Palm, this PR [openedx/edx-platform#32350](https://github.com/openedx/edx-platform/pull/32350) has upgraded Django storages, so if you use any platform version since June 5, you'll get this error. Ultimately though as I said, I think the bug is an upstream bug so I'm going to open a PR to fix it upstream.
> 
> Edit: And it turns out that @ormsbee already beat me to the upstream fix :) [openedx/edx-platform#32733](https://github.com/openedx/edx-platform/pull/32733)

Thanks for looking into this! I feel like given that warning, perhaps it is best to add the setting you recommended anyway? It is a pretty small fix, so I've added it to the PR, and will kick off another build to see if all works well.

**Comment by fghaas** [comment]
Excellent, thanks! @mrtmm, please give this PR an extra pair of eyes, just to make sure I didn't miss anything important. :)

**Review by mrtmm — APPROVED** [review]
Looks good to me, thanks!

**Comment by fghaas** [comment]
Merged as-is.

@bradenmacdonald, re whether the `COURSE_METADATA_EXPORT_STORAGE` fix is still necessary, I'm a little confused. Maybe you can shed some light on this:

* https://github.com/openedx/edx-platform/pull/32350 landed in `master` on 2023-06-05. It doesn't look to me as if it was ever cherry-picked into `open-release/palm.master`.
* Although `open-release/palm.1` [was tagged a week later on 2023-06-12](https://github.com/openedx/edx-platform/tags), because of that missing cherry-pick it never upgraded to django-storages 1.9.
* https://github.com/openedx/edx-platform/pull/32733 landed on 2023-07-17. It _also_ hasn't been cherry-picked into  `open-release/palm.master`.

As a consequence, neither of these changes apply to someone running `open-release/palm.1` (which is [what Tutor installs by default](https://github.com/overhangio/tutor/blob/3affd546e9fcdacb09d5c579a6721671eea064c9/tutor/templates/config/defaults.yml#L55) at the time of writing).

So, effectively _all of Palm_ is currently setting `COURSE_METADATA_EXPORT_STORAGE` incorrectly, and although post-Palm `master` has fixed that issue, it's not really clear to me if that fix will ever land in Palm at all, or whether it won't (and presumably, Palm will stay on django-storages 1.8 throughout its release cycle). Do you happen to know more?

**Comment by bradenmacdonald** [comment]
@fghaas I think you have everything correct, and I agree with you and @xitij2000 that for now we might as well include the override that you put in https://github.com/hastexo/tutor-contrib-s3/pull/63 just to be on the safe side. There's no harm in doing it, and it will silence the warning and help prevent a future bug if the version bump gets backported to Palm at some point.

**Comment by fghaas** [comment]
@bradenmacdonald ack, thanks! I guess we'll go ahead with #63 then.

**Pull request merged**